### PR TITLE
🏗 ♻️ Fix type errors in /build-system/tasks/e2e

### PR DESCRIPTION
**Pull request by rileyajones**
/build-system/tasks/e2e is directory which contains the largest number of type errors of any of build-system tasks. This PR is intended to be non functional and should resolve all simple type errors within the subdirectory. After this lands there will still be a handful of non trivial errors requiring functional changes that will be fixed as part of another PR.

#28387

### CODE REVIEW

File: build-system/tasks/e2e/selenium-webdriver-controller.js [hunk 1]
```diff
@@ -13,18 +13,21 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+// @ts-nocheck
```
Comment (estherkim) [inline build-system/tasks/e2e/selenium-webdriver-controller.js:16]: why? is this a particularly bad one?

File: build-system/tasks/e2e/selenium-webdriver-controller.js [hunk 2]
```diff
@@ -144,15 +146,15 @@ class SeleniumWebDriverController {
...
+    return new ElementHandle(webElement);
   }
 
   /**
    * This waits for elements to become available similar to Selenium's
    * until.js#elementsLocated
    * {@link https://github.com/SeleniumHQ/selenium/blob/6a717f20/javascript/node/selenium-webdriver/lib/until.js#L258}   *
    * @param {string} selector
-   * @return {!Promise<!Array<!ElementHandle<!WebElement>>>}
+   * @return {!Promise<!Array<!ElementHandle<!selenium.selenium.WebElement>>>}
```
Comment (estherkim) [inline build-system/tasks/e2e/selenium-webdriver-controller.js:157]: selenium.selenium ?

File: build-system/tsconfig.json [hunk 3]
```diff
@@ -27,15 +27,18 @@
     "./common/**/*.ts",
     "./server/**/*.js",
     "./server/**/*.ts",
-    "./tasks/**/*.js",
-    "./tasks/**/*.ts"
+    "./tasks/e2e/*.js",
```
Comment (estherkim) [inline build-system/tsconfig.json:30]: same comment about combining *.ts and *.js lines.

File: build-system/tasks/e2e/selenium-webdriver-controller.js (hunk 1 above)
Comment (rileyajones) [reply build-system/tasks/e2e/selenium-webdriver-controller.js:16]: Yeah, the `ControllerPromise` type is annotated as though it extends `Promise` and treated as such but does not actually extend `Promise`. I was hoping to address that as a functional change in a separate PR.

File: build-system/tsconfig.json (hunk 3 above)
Comment (rileyajones) [reply build-system/tsconfig.json:30]: Pretty sure I wrote this prior to your other comment and didn't remember to make the change on this branch. Great suggestion as before.

File: build-system/tasks/e2e/selenium-webdriver-controller.js (hunk 2 above)
Comment (rileyajones) [reply build-system/tasks/e2e/selenium-webdriver-controller.js:157]: ☠️ CMD+OPTION+F

File: build-system/tasks/e2e/functional-test-controller.js [hunk 4]
```diff
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+// @ts-nocheck
```
Comment (rsimha) [inline build-system/tasks/e2e/functional-test-controller.js:16]: Is this intentional? If yes, maybe a better option is to exclude this file via `build-system/tsconfig.json`?

File: build-system/tasks/e2e/puppeteer-controller.js [hunk 5]
```diff
@@ -30,6 +25,13 @@ const {
 const {ControllerPromise} = require('./controller-promise');
 const {dirname, join} = require('path');
 
+const {
+  Browser, // eslint-disable-line no-unused-vars
+  JSHandle, // eslint-disable-line no-unused-vars
+  Page, // eslint-disable-line no-unused-vars
+  ElementHandle: PuppeteerHandle, // eslint-disable-line no-unused-vars
```
Comment (rsimha) [inline build-system/tasks/e2e/puppeteer-controller.js:32]: Is there a way to avoid these?

File: build-system/tasks/e2e/selenium-webdriver-controller.js (hunk 1 above)
Comment (rsimha) [reply build-system/tasks/e2e/selenium-webdriver-controller.js:16]: Perhaps this file too should be excluded via `build-system/tsconfig.json`?

File: build-system/tsconfig.json [hunk 6]
```diff
@@ -29,8 +29,8 @@
     "./tasks/coverage-map/*",
     "./tasks/create-golden-css/*",
     "./tasks/css/*",
-    // "./tasks/e2e/*", // 253 remaining type errors.
-    "./tasks/make-extension/*",
+    "./tasks/e2e/*", // 253 remaining type errors.
+    "./tasks/extension-generator/*",
```
Comment (rsimha) [inline build-system/tsconfig.json:33]: Why this change? (`extension-generator` no longer exists.)

**Review by rsimha — APPROVED** [review]
Getting closer to zero errors! LGTM with a few more comments. Okay to merge after they are addressed and CI is green.

File: build-system/tasks/e2e/functional-test-controller.js (hunk 4 above)
Comment (rileyajones) [reply build-system/tasks/e2e/functional-test-controller.js:16]: Unfortunately that won't work because the file is imported by other files which are in the `included` list. (It's not very intuitive behavior) https://github.com/microsoft/TypeScript/issues/36017 

That being said in the time since this PR was written and now the number of issues in he build-system directory has ballooned again so I think I'll just remove this for now and try to fix the root issue in another PR.

File: build-system/tasks/e2e/puppeteer-controller.js (hunk 5 above)
Comment (rileyajones) [reply build-system/tasks/e2e/puppeteer-controller.js:32]: I could condense them, or else I could try to do some dynamic typing or just drop type enforcement.

File: build-system/tasks/e2e/selenium-webdriver-controller.js (hunk 1 above)
Comment (rileyajones) [reply build-system/tasks/e2e/selenium-webdriver-controller.js:16]: See https://github.com/ampproject/amphtml/pull/33166#discussion_r612793826. This file "only" is responsible for 9 errors but they are pretty hard (maybe impossible?) to fix in a non functional way. I'll remove this ts-nocheck for now for the same reasons stated in the linked comment and will address this file in its own PR.

File: build-system/tsconfig.json (hunk 6 above)
Comment (rileyajones) [reply build-system/tsconfig.json:33]: Good catch, I suppose this is the issue with resurrecting old PRs.
image

File: build-system/tasks/e2e/controller-promise.js [hunk 7]
```diff
@@ -80,14 +79,18 @@ class ControllerPromise {
...
+ * @template VALUE
+ * @template RES
  */
 function wrapWait(wait, mutate) {
   return (condition, opt_mutate) => {
-    opt_mutate = opt_mutate || ((x) => x);
+    opt_mutate = opt_mutate || ((x) => /** @type {*} */ (x));
     return wait(condition, (value) => opt_mutate(mutate(value)));
   };
 }
```
Comment (rileyajones) [inline build-system/tasks/e2e/controller-promise.js:96]: This was a heck of a puzzle. More than open to changing type names etc...

Comment (estherkim) [reply build-system/tasks/e2e/controller-promise.js:96]: Go for it!

**Review by estherkim — APPROVED** [review]
